Automattic/woocommerce-payments
# Add Playwright e2e test suite ready for incremental migration and visual regression testing

Resolves #8177

#### Changes proposed in this Pull Request

This PR lays the foundation for Playwright e2e and visual regression testing, ready for incremental migration of existing Puppeteer e2e tests.

Migrating our e2e tests to use Playwright instead of Puppeteer will provide us benefits including, but not limited to:
* Reduced test flakiness/increased test stability with [auto-waiting](https://playwright.dev/docs/actionability) and configurable test [retries](https://playwright.dev/docs/test-retries).
* Easier to write robust and maintainable e2e tests with [locators](https://playwright.dev/docs/locators#quick-guide) for targeting DOM elements using testing-library APIs (consistent with integration test APIs).
* [Visual regression testing](https://playwright.dev/docs/test-snapshots) to help us update component library dependencies with more confidence.


To aid in migration, this PR follows the [WooCommerce Puppeteer to Playwright migration guide](https://github.com/woocommerce/woocommerce/wiki/Puppeteer-to-Playwright-Migration-Guide-for-e2e-Testing) which adds a Playwright test suite alongside the existing Puppeteer test suite.

**TODO**

- [x] Migrate example tests
- [x] Setup GH check
- [x] Add basic tips on how to write PlayWright tests in `tests/e2e-pw/README.md`
- [x] #8160 
- [x] Review and merge this PR 🚀
- [x] Discuss handover with Harmony – FE guild will remain caretakers until a future time TBD
- [x] Post the draft P2 post (paJDYF-cjc-p2) to inform stakeholder teams and encourage adoption (link to `tests/e2e-pw/README.md` for getting started)


**Notes/inclusions:**

* The fundamental Playwright boilerplate and config.
* Authentication setup step for admin & customer with caching for local dev envs.
* Some initial basic tests with visual regression snapshots.
* Uses the existing e2e docker environment setup.
* Uses many of the same conventions from the [WC Puppeteer to Playwright migration guide](https://github.com/woocommerce/woocommerce/wiki/Puppeteer-to-Playwright-Migration-Guide-for-e2e-Testing) (e.g. `e2e-pw` directory).
* A GH action that runs e2e-pw on PRs to `develop` or `trunk` and outputs the Playwright HTML report as a GH action artefact ([see this workflow run](https://github.com/Automattic/woocommerce-payments/actions/runs/7218882881?pr=7922))

<img width="1038" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/ce1d4bb1-d69c-4e1a-afc5-5185388873b7">


#### Testing instructions
1. `npm install`
1. Setup e2e environment
	1. `npm run test:e2e-setup`
	1. `npm run test:e2e-up`
1. Run e2e tests with Playwright
	-   `npm run test:e2e-pw` headless run from within a linux docker container.
	-   `npm run test:e2e-pw-ui` runs tests in interactive UI mode from within a linux docker container – recommended for authoring tests and re-running failed tests.
	-   `npm run test:e2e-pw keyword` runs tests only with a specific keyword in the file name, e.g. `dispute` or `checkout`.
	-   `npm run test:e2e-pw --update-snapshots` updates snapshots.
1. Ensure the GH check `E2E Playwright Tests` passes
	- View the summary of the GH check run and download the `playwright-report.zip`
	- Open the `index.html` locally in your browser to view the test report
1. Ensure Puppeteer tests pass

-------------------


**Post merge**
